app/code/Magento/CatalogGraphQl/DataProvider/Product/SearchCriteriaBuilder.php

```diff
@@ -122,7 +122,7 @@ public function build(array $args, bool $includeAggregation): SearchCriteriaInte
         }
 
         $this->addEntityIdSort($searchCriteria);
-        $this->addVisibilityFilter($searchCriteria, $isSearch, !empty($args['filter']));
```

Comment:
I think this logic doesn't make sense: https://github.com/magento/magento2/blob/688a0b4172ffea536628b0d22fbdb833cb86e33f/app/code/Magento/CatalogGraphQl/DataProvider/Product/SearchCriteriaBuilder.php#L142

If a search query is present it should always use `getVisibleInSearchIds` and if it doesn't contain a search query it should use `getVisibleInCatalogIds`. I don't think it makes sense to not filter anything. We are either in 'search mode' or in 'catalog mode' not in both?

Reply:
@paales , Which you mentioned this logic `if ($isFilter && $isSearch) { ` , is not related to this PR changes, it was available from core standards.
This PR is related to filter category conditions for catalog visibility combination  `$this->addVisibilityFilter($searchCriteria, $isSearch, !empty($args['filter']['category_id']));`

Reply:
I don't think selecting by category_id is a valid condition:
- When searching for a keyword but _filtering_ on a category_id it should still have visibility search? Or you could argue that each item should be in Search AND Catalog, but not neither? 
- You only check on `category_id` but we've also got `category_uid` so, that should probably be changed.

Reply:
@paales I have checked in the codebase this `build` method has been called from the below file

https://github.com/magento/magento2/blob/cfaeeaef415c2a77009db57995e5eee627350878/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php#L185

And if from the query, someone is passing `category_uid` then it is converted in `category_id` through the below function call:

https://github.com/magento/magento2/blob/cfaeeaef415c2a77009db57995e5eee627350878/app/code/Magento/CatalogGraphQl/Model/Resolver/Products/Query/Search.php#L190